### PR TITLE
Versioning, hoisting and fixed links

### PR DESCRIPTION
**Pull request by arobson**
Instead of fussing with PR order, I just put everything into one honking PR. You'll want to get autohost's PRs merged and published before merging this one. Includes a fix for link rendering, action hoisting and handle versioning.

### CODE REVIEW

File: spec/integration/autohost.spec.js [hunk 1]
```diff
@@ -101,6 +102,26 @@ describe( "Autohost Integration", function() {
 			} );
 		} );
 
+		describe( "when requesting intermediate board json version", function() {
```
Comment (digitalBush) [inline spec/integration/autohost.spec.js]: Would be good to see a test case for version 10 too.

File: src/versions.js [hunk 2]
```diff
@@ -48,7 +61,38 @@ function getVersions( resource ) { // jshint ignore:line
...
 
+function processHandles( resource ) {
+	if ( resource.versions ) {
+		_.each( resource.versions, function( change, version ) {
+			version = parseInt( version );
+			_.each( change, function( action, name ) {
+				if ( action.handle ) {
+					var targetAction = resource.actions[ name ];
+					if ( !_.isArray( targetAction.handle ) ) {
+						var original = targetAction.handle;
```
Comment (digitalBush) [inline src/versions.js]: Can we introduce a new endpoint at a specific version?

**Comment by digitalBush** [comment]
Locally I'm seeing 3 failed tests:

```
  1) Autohost Integration with newest version as default when requesting board with no media type should get JSON version 2:

      AssertionError: expected {} to deeply equal { Object (id, title, ...) }
      + expected - actual

      -{}
      +{ /* lots of stuff */}

      at Context.<anonymous> (spec/integration/autohost.spec.js:378:17)

  2) Autohost Integration with undefined api prefix when requesting board with no media type should get JSON version 2:

      AssertionError: expected {} to deeply equal { Object (id, title, ...) }
      + expected - actual

      -{}
      +{ /* lots of stuff */}

      at Context.<anonymous> (spec/integration/autohost.spec.js:473:17)

  3) Autohost Integration with empty api prefix when requesting board with no media type should get JSON version 2:

      AssertionError: expected {} to deeply equal { Object (id, title, ...) }
      + expected - actual

      -{}
      +{ /* lots of stuff */}
```

**Comment by arobson** [comment]
@digitalBush Q: what do I do when I wake up at 3 am and can't get back to sleep? A: fix my crap.

So this thing was 1 part me being dumb, 1 part lodash being a bit of a jerk and 1 part different contexts causing iteration to happen in different orders. (i.e. I couldn't reproduce your issue until much later and only then after having put code in to enforce sorting).

You'll notice the versions module got a decent overhaul, that's where most of the problems were occurring. There was something odd happening where not cloning the source property values into the target version was creating some really odd behavior. There was that whole version iteration not happening in the order one might expect, so now version does weird gymnastics to enforce order when applying change sets to get a desired version. Lastly, there was a "fun" bug because of how I was using lodash that resulted in array properties from a new version not over-writing the original version if the newer array had fewer elements.

All of this _nearly_ convinces me that I'm very bad at this "programming" thing.

File: src/index.js [hunk 3]
```diff
@@ -165,6 +170,10 @@ function hyperMiddleware( state, req, res, next ) { // jshint ignore:line
 	var hyperModel = getHyperModel( state, req );
 	var envelope = getEnvelope( req, res );
 	var response = new HyperResponse( envelope, engine, hyperModel, contentType ).origin( req.originalUrl, req.method ); // jshint ignore:line
+	if ( !req.context ) {
+		req.context = {};
+	}
+	req.context.version = parseInt( getVersion( state, envelope ) );
```
Comment (digitalBush) [inline src/index.js]: Any reason by getVersion would return something besides an int?

File: spec/integration/autohost.spec.js [hunk 4]
```diff
@@ -352,7 +394,7 @@ describe( "Autohost Integration", function() {
 				} );
 			} );
 
-			it( "should get JSON version 2", function() {
+			it( "should get version 10's hot nonsense", function() {
```
Comment (digitalBush) [inline spec/integration/autohost.spec.js]: :joy:

File: spec/integration/halBoard.json [hunk 5]
```diff
@@ -65,7 +67,8 @@
 									"href": "/test/api/card/301/block",
 									"method": "PUT"
 								}
-							}
+							},
+							"_version": 1
```
Comment (digitalBush) [inline spec/integration/halBoard.json]: Curious, would this version ever be different than the parent?

File: src/hyperResource.js [hunk 6]
```diff
@@ -313,14 +317,35 @@ function resourceGenerator( state, envelope, data, parentUrl, originUrl, originM
...
+								link[ hoistedActionName ] = link[ actionName ];
+								delete link[ actionName ];
+								return link;
+							} );
+					}
+				} );
+			} );
+		}
+
+		return when.all( ownLinks.concat( _.flatten( otherLinks ) ) )
```
Comment (digitalBush) [inline src/hyperResource.js]: You can do a `flatMap` on 332 and skip the `flatten` step here I believe.

File: src/versions.js [hunk 7]
```diff
@@ -36,7 +63,17 @@ function getVersion( resource, version ) { // jshint ignore:line
 	if ( version === undefined ) {
 		return resource;
 	} else {
-		return getVersions( resource )[ version ] || resource;
+		var versions = getVersions( resource );
+		version = parseInt( version );
```
Comment (digitalBush) [inline src/versions.js]: Calvin pointed out the `radix` parameter. Apparently the [default behavior](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt) will treat the value as octal if it starts with a "0". :tired_face:

File: spec/integration/halBoard.json (hunk 5 above)
Comment (arobson) [inline spec/integration/halBoard.json]: It won't. You can't request the parent resource as a version and get something other than that from the child. Even if the child didn't have version X, the way versions are applied, the child will have the same "effective" version despite there being no difference between its last version and the parent's.

File: src/versions.js (hunk 7 above)
Comment (arobson) [inline src/versions.js]: That's ok. You also can't prefix versions with "psketti" or a hex bitmap of Donald Trump. Unless there's some RFC or examples out there of people prefixing version numbers with 0s, I'mma just leave this be :grin:

Comment (digitalBush) [inline src/versions.js]: :+1:

File: spec/integration/halBoard.json (hunk 5 above)
Comment (digitalBush) [inline spec/integration/halBoard.json]: If that's the case, do we need to include `_version` on all of the embedded resources?

**Comment by arobson** [comment]
@digitalBush thanks for suffering through this one and helping make it moar gooder :clap:

**Pull request merged**